dali/dfuplus/dfuplus.cpp

```diff
@@ -1448,7 +1452,262 @@ int CDfuPlusHelper::abort()
     {
         error("Aborted %s\n", wuid);
     }
-    
+
+    return 0;
+}
+
+int CDfuPlusHelper::listhistory()
+{
+    const char* lfn = globals->queryProp("lfn");
+    if(lfn == NULL || *lfn == '\0')
+            throw MakeStringException(-1, "srcname not specified");
+
+    Owned<IClientListHistoryRequest> req = dfuclient->createListHistoryRequest();
+    req->setName(lfn);
+
+    Owned<IClientListHistoryResponse> resp = dfuclient->ListHistory(req);
+
+    const IMultiException* excep = &resp->getExceptions();
+    if(excep != NULL && excep->ordinality() > 0)
+    {
+        StringBuffer errmsg;
+        excep->errorMessage(errmsg);
+        error("%s\n", errmsg.str());
+        return -1;
+    }
+
+    typedef enum
+    {
+        xml = 0,
+        ascii = 1,
+        csv = 2,
+        json = 3,
+    }out_t;
+
+    out_t format;
+
+    const char *formatPar = globals->queryProp("outformat");
+    if(formatPar == nullptr)
+        format = xml;
+    else if (stricmp(formatPar, "csv") == 0)
+        format=csv;
+    else if (stricmp(formatPar, "xml") == 0)
+        format=xml;
+    else if (stricmp(formatPar, "json") == 0)
+        format=json;
+    else if (stricmp(formatPar, "ascii") == 0)
+        format=ascii;
+    else
+    {
+        error("Unsupported output format: '%s'\n",formatPar);
+        return -1;
+    }
+
+    MemoryBuffer tmp;
+    tmp.append(resp->getXmlmap());
+    if (0 == *tmp.toByteArray())
```

Comment:
The problem and why I did that magic is:
/home/ati/HPCC-Platform/dali/dfuplus/dfuplus.cpp:1506:44: error: binding ‘const MemoryBuffer’ to reference of type ‘MemoryBuffer&’ discards qualifiers
     MemoryBuffer &respMsg = resp->getXmlmap();
                                            ^
Moreover (and reflecting your further comment) if I change respMsg to const reference and  apply further code as you suggested:
    const MemoryBuffer &respMsg = resp->getXmlmap();
    if (0 == respMsg.length())
    {
        error("%s doesn't have stored history!\n", lfn);
        return -2;
    }
    Owned<IPropertyTree> history;
    history.setown(createPTree(respMsg));

I can get an error again:
/home/ati/HPCC-Platform/dali/dfuplus/dfuplus.cpp:1516:39: error: binding ‘const MemoryBuffer’ to reference of type ‘MemoryBuffer&’ discards qualifiers
     history.setown(createPTree(respMsg));

Based on the generated virtual const MemoryBuffer & getXmlmap()=0;

Is there any, easier way to "convert"  const MemoryBuffer & to MemoryBuffer & ?